django/thunderstore/repository/models/package_version.py

```diff
@@ -297,6 +324,116 @@ def log_download_event(version_id: int, client_ip: Optional[str]):
 
         log_version_download.delay(version_id, timezone.now().isoformat())
 
+    def build_audit_event(
+        self,
+        *,
+        action: AuditAction,
+        user_id: Optional[int],
+        message: Optional[str] = None,
+    ) -> AuditEvent:
+        return AuditEvent(
+            timestamp=timezone.now(),
+            user_id=user_id,
+            action=action,
+            message=message,
+            related_url=self.package.get_view_on_site_url(),
+            fields=[
+                AuditEventField(
+                    name="Package",
+                    value=self.package.full_package_name,
+                ),
+            ],
+        )
+
+    @transaction.atomic
+    def reject(
+        self,
+        agent: Optional[UserType],
+        is_system: bool = False,
+        message: Optional[str] = None,
+    ):
+        if self.review_status == PackageVersionReviewStatus.immune:
+            raise PermissionError()
+
+        if is_system or self.can_user_manage_approval_status(agent):
+            self.review_status = PackageVersionReviewStatus.rejected
+            self.save(update_fields=("review_status",))
+
+            fire_audit_event(
+                self.build_audit_event(
+                    action=AuditAction.VERSION_REJECTED,
+                    user_id=agent.pk if agent else None,
+                    message=message,
+                )
+            )
+        else:
+            raise PermissionError()
+
+    @transaction.atomic
+    def approve(
+        self,
+        agent: Optional[UserType],
+        is_system: bool = False,
+        message: Optional[str] = None,
+    ):
+        if self.review_status == PackageVersionReviewStatus.immune:
+            raise PermissionError()
+
+        if is_system or self.can_user_manage_approval_status(agent):
+            self.review_status = PackageVersionReviewStatus.approved
+            self.save(update_fields=("review_status",))
+
+            fire_audit_event(
+                self.build_audit_event(
+                    action=AuditAction.VERSION_APPROVED,
+                    user_id=agent.pk if agent else None,
+                    message=message,
+                )
+            )
+        else:
+            raise PermissionError()
+
+    def can_user_manage_approval_status(self, user: Optional[UserType]) -> bool:
+        if self.review_status == PackageVersionReviewStatus.immune:
+            return False
+
+        for listing in self.package.community_listings.all():
+            if listing.can_user_manage_approval_status(user):
+                return True
+        return False
+
+    @transaction.atomic
+    def update_visibility(self):
+        if not self.visibility:
+            self.visibility = VisibilityFlags.objects.create_unpublished()
```

Comment:
I suspect this should be an impossible codepath if the mixin is being used appropriately & can be removed